### ISSUE TITLE
Set up a JupyterHub that can run CUDA code for NeurIPS

### ISSUE THREAD

**Issue by jzf2101** [jupyterhub/team-compass#92]
We need to set up at JupyterHub with CUDA and then turn it into a binderhub 

See https://github.com/jupyterhub/team-compass/issues/52 and https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/994#issue-373992464

@choldgraf @betatim @aculich @consideRatio @cmd-ntrf 

Based on #81 we could use the binder account?

**Comment by choldgraf** [comment]
yes, feel free to use the binder account, but please be mindful of the $$$ you burn through :-)

**Comment by consideRatio** [comment]
I'm happy to do the initial setup of the GKE cluster and node pools etc, but I would need to be added to the google cloud project it should be done in if that is alright.

**Comment by cmd-ntrf** [comment]
Correct me if I am wrong, but unless pods have the ability to share a GPU in GKE, we will face issue when running more containers than we have GPUs. This potentially limits the scalability of Binder in these circumstances. 

I guess as long as it is only for a demo, this is fine, but if we want people to try it on their own, we might need to setup a generous cluster.

**Comment by jzf2101** [comment]
@cmd-ntrf based on #81 we are planning on limiting the scaling to put a ceiling on the number of concurrent users

**Comment by consideRatio** [comment]
@jzf2102, the cheapest gpus will do right? they are still VERY powerful GPUs (nvidia tesla k80).

I need to tie the GKE cluster to a region and somewhat more loosly to a zone/datacenter and each zone has only certain GPUs. Ill choose a zone where k80s and p100s are available, as compared to one with v100 only are available perhaps (k80 < p100 < v100 < TPUs)

**Comment by jzf2101** [comment]
Yeah I figured that we could give people 1 K80 ideally since the demo is in eastern Canada somewhere in eastern Canada or easteern US would work? Perhaps we should give them the option to use a K80 or not so we don't have to use a K80 every time.

**Comment by choldgraf** [comment]
Minimal Viable Product is totally fine in this case, IMO. We're just giving people a proof-of-concept here, don't need to mine bitcoin for the whole conference :-)

**Comment by betatim** [comment]
@consideRatio nice to have you helping out :) Let me know if you need to know something about the current setup in the GKE project.

The cost calculation we did was based on having one `n1-standard-2` instance per pod and giving each instance a K80. As we will not auto-scale the cluster I think having one node per pod (or user) with one GPU attached to it will simplify things for the demo.

(I think in the long term (and "at scale") we'd want a setup where the node your notebook runs on doesn't actually have all that many resources and instead you use a library/tool to create worker Job(s) that run on a GPU enabled instance. Similar to how kubeflow works and dask. But this is something to discuss after the demo.)

**Comment by consideRatio** [comment]
@betatim ah i appreciate the insight about kubeflow, really wanted to learn more but have not done it yet. I know there is work done at making request to fractions of GPU, and ive heard timeframes of 6 months, but im thinking 12 months in practice. I think it will be best to use scheduled jobs for the gpu though no matter what as the workload tend to be bursty for ML purposes, and the GPU memory usage may peak hard during usage so using request/limits may be too simple solution to a complex problem anyhow.

@jzf2101: k80 gpu, eastern us, profile_list with cpu and gpu option provided.

*Who* points *what domain* to the IP later btw?

**Comment by choldgraf** [comment]
we can point a sub-domain of mybinder.org to the deployment (e.g. neurips.mybinder.org and hub.neurips.mybinder.org)

**Comment by jzf2101** [comment]
We don't need that for initial setup though correct?

**Comment by consideRatio** [comment]
@jzf2101 nope no need for that during the initial setup

**Comment by consideRatio** [comment]
Creating a *regional cluster* named neurips in the binder-prod google project. It will have nodes only on the zone `us-east1-d`. The only option would been `us-east1-c` assuming we want `us-east` and a zone with NVIDIA Tesla K80 GPU nodes.

(It is a good practice to always go for regional clusters and native VPC according to google k8s experts.)

**Comment by jzf2101** [comment]
So wait `us-east1-d` doesn't have K80s?

**Comment by betatim** [comment]
Is there a repository already that has the configuration/deploy/helm chart in it?

**Comment by jzf2101** [comment]
Status update- we have a binder hub up now and I can start putting images on it, but if we don't have k80s associated how doe we know that the repos I put in with k80s work?

**Comment by jzf2101** [comment]
@choldgraf I think there's been chatter on the gitter channel about what's going on.  From what I recall, I think we're waiting for an increase in quota from GCP.  @consideRatio has requested it but we haven't gotten it yet. We have 1 K80 and 100 preemptible ones - I'm assuming they're also K80s?

**Comment by consideRatio** [comment]
We now have 20 k80 GPU standard GPUs, and 100 preemptible ones.
Work still remains, and I'll spend time in between 19-01 tonight Swedish time when I'm off work.

**Comment by betatim** [comment]
This comment serves as a todo list for the deployment (for repositories to run on the deployment use a different issue/comment). The list is sorted (top to bottom) from "absolutely needed there will be no demo without this" to "nice to have" to "let's try this out for extra credit". If you have a moment to contribute please work on things that are at the top of the list, when you check them off edit this post to add a link to where you performed that change, don't tackle things further down if there are open items towards the top.

---

Currently deployed: https://github.com/consideRatio/neurips.mybinder.org-deploy/commit/d7ab128f6da4d3a956c12571dde55acf7de0765e
Current GPU VM pool size: 0

---

To do list (please edit it if you think of things that need to get done, keeping the sort order in mind, so insert them at a good place, not at the end):
* [x] basic deployment with GPU support works (https://github.com/consideRatio/neurips.mybinder.org-deploy/commit/d7ab128f6da4d3a956c12571dde55acf7de0765e)
* [x] demo repository to verify GPU support (https://github.com/betatim/basic-gpu-binder)
* [x] request K80 quota increase to 100 in us-east1
* [ ] create repository that runs a well known notebook from the keras/tensorflow/pytorch documentatation
* [ ] configure and check inactive pods are culled within 5minutes (more aggressive than on mybinder.org)
* [ ] configure JupyterHub to emit an error message when max number of pods is reached ([here](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/f202306ad49e3883fc55ab89cdb41f64b87c8c86/jupyterhub/values.yaml#L17)) Should be set to match cluster size
* [ ] establish protocol around manually scaling cluster up/down: who does it, where to post what the limit should be, how to do it
* [ ] limit outgoing network traffic from user pods
* [ ] make sure CORS is disabled so that neurips.mybinder.org can't be used by nbinteract/thebe/etc

**Comment by minrk** [comment]
CORS should be disabled by default, but it's worth verifying.

**Comment by minrk** [comment]
5 minutes might be too aggressive to cull. The activity metric is only updated on the Hub periodically when checking in with the proxy, so an acitivity

> Utilize JupyterLab by default with the BinderHub

Setting `c.Spawner.default_url = '/lab/` should be all that's needed for this, I think.

**Comment by consideRatio** [comment]
Would it function with provided "start this file on startup" as well?

**Comment by minrk** [comment]
Ah, no. You'd still need to use 'url to open' instead of 'file to open'

**Comment by minrk** [comment]
.We can teach binderhub about jupyterlab's file-open urls now that they exist. Then we can have an option on binderhub to build `/lab/tree/...` urls instead of `/notebooks/...`.

**Comment by consideRatio** [comment]
@betatim @minrk should we go for lab or not even without the fileopen support?

**Comment by minrk** [comment]
I wouldn't do it today, no.

**Comment by consideRatio** [comment]
# Deployment oversight guide
The conference is in Montreal (GMT-5), and the actual demonstration starts at 4 PM.
This is the planned procedure.

## Monday night, current
We now have 5 GPU available and a single user-placeholder
## Tuesday morning
#### At 8 AM @consideRatio does the following ( 14:00 Swedish time )
  - Scale placeholders to 5
  ```
  kubectl scale -n neurips sts/user-placeholder --replicas 5
  ```
  - Update the autoscaler's upper limit to 100 [in GCP](https://console.cloud.google.com/kubernetes/clusters/details/us-east1/neurips?project=binder-prod)
#### At 3 PM @choldgraf does the following, and @consideRatio will make sure its ok ( 21:00 Swedish time )
- Scale placeholders to 15
  ```
  kubectl scale -n neurips sts/user-placeholder --replicas 5
  ```
#### Later @choldgraf may decide to lower the placeholders, or just let it be as it is
- Scale placeholders to X
  ```
  kubectl scale -n neurips sts/user-placeholder --replicas X
  ```
## Thursday evening
3. When its time to shut it all down, we do the following procedure: 
  ```
  # remove the placeholders
  kubectl scale -n neurips sts/user-placeholder --replicas 0
  ```
  ```
  # remove the actual users
  kubectl delete pod -n neurips -l component=singleuser-server
  ```
  - Update the autoscaler's upper limit to 0 [in GCP](https://console.cloud.google.com/kubernetes/clusters/details/us-east1/neurips?project=binder-prod)
  - Force resize the cluster to 0 nodes [in GCP](https://console.cloud.google.com/kubernetes/clusters/details/us-east1/neurips?project=binder-prod)

**Comment by betatim** [comment]
This is a great plan/schedule!

For the point on pre-pulling images: I think you mean the right thing but the way it is written is ambiguous so my two cents: you need to look at the name of the image that repo2docker creates for that particular repository. This is different from the name of the pod/image used by any pod running. There are two ways to see the image name:
* look in the registry for the image and tag after the build has finished
* `kubectl describe` the `jupyter-...` pod that gets launched after the build succeeds

**Comment by consideRatio** [comment]
ah yepp, i was thinking of the `-o yaml` as well, in practice:
```
kubectl get pod -o yaml <podname> | grep "image: gcr"
```
